Updates for QueryClient HowTo notebook

jacquesalice opened the pull request.

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -40,36 +45,52 @@
+    "The status of an asynchronous query can be checked with *queryClient.status(jobid)*. A return value of 'COMPLETED' indicates the query has terminated. A return value of 'ERROR' indicates that there was a problem with the query.    \n",
     "\n",
-    "The results of an asynchronous query (assuming that they were not saved to either the user's virtual storage or remote database) can be retrieved once the query has completed with an HTTP GET call to the query manager service <i>results</i> endpoint with the relevant job id as argument: /results?jobid=<jobid>\n",
+    "The results of an asynchronous query (assuming that they were not saved to either the user's virtual storage or remote database) can be retrieved once the query has completed with *result = qc.query(adql=query,async_=True,wait=True,poll=1,verbose=1)*\n",
```
I propose to let this be two cases, both described:

1) a manual retrieval using a jobid:

jobid = qc.query(sql=query,async+=True)
print(jobid)
print(qc.status(jobid)) # repeat until 'COMPLETED'
result = qc.results(jobid)

2) A mechanism to run an async query but to make the client auto-loop until the result is available:

result = qc.query(adql=query,async_=True,wait=True,poll=1,verbose=1)

Only case 2) should be coded in the notebook. (but case 1 can be linked to in the 
01_GettingStartedWithDataLab/01_JupyterPythonSQL101.ipynb NB, or even repeated here).

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -40,36 +45,52 @@
     "\n",
     "#### Output formats\n",
-    "The results can be returned as whitespace delimited (<i>ascii</i>), CSV (<i>csv</i>), FITS object (<i>fits</i>), HDF5 (<i>hdf5</i>, or in VOTable format (<i>votable</i>). Note that if the results are saved to the user's database then the output format is ignored.\n",
+    "The results can be returned as whitespace delimited (*ascii*), CSV (*csv*), FITS object (*fits*), HDF5 (*hdf5*, or in VOTable format (*votable*). Note that if the results are saved to the user's database then the output format is ignored.\n",
```
hd5 needs a closing parenthesis

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -40,36 +45,52 @@
     "### From Python code\n",
     "\n",
-    "The query client service can be called from Python code using the <i>datalab</i> module. This provides methods to access the various query manager functions in the <i>queryMgr</i> subpackage. See the information [here](https://github.com/noaodatalab/datalab/blob/master/README.md).\n"
+    "The query client service can be called from Python code using the *datalab* module. This provides methods to access the various query client functions in the *QueryClient* subpackage. See the information [here](https://github.com/noaodatalab/datalab/blob/master/README.md).\n"
```
Suggest "Queries can be also run from the command line, e.g. on your local machine, using the `datalab` command line utility. It provides methods [...]." And then link straight to the GitHub repo, i.e. https://github.com/noaodatalab/datalab

jacquesalice replied:
The datalab command line notebook hasn't been posted yet but should it link to that instead?

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -40,36 +45,52 @@
+    "\n",
+    "* Data Lab concept paper: Fitzpatrick et al., \"The NOAO Data Laboratory: a conceptual overview\", SPIE, 9149, 2014, http://dx.doi.org/10.1117/12.2057445\n",
+    "\n",
+    "* Data Lab disclaimer: http://datalab.noao.edu/disclaimers.php"
```
Change http to https

jacquesalice replied:
I will change it here, but it links to the same site. And http is used on all of our notebooks, it is http in our template.

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
````diff
@@ -99,54 +120,57 @@
+    "WHERE constraints\n",
+    "LIMIT 100\n",
+    "```\n",
+    "Please see our intro notebook *01_JupyterPythonSQL101* for more info on this."
````
Link to the NB directly (on Github)

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -99,54 +120,57 @@
-    "### A quick query\n",
+    "# A quick query\n",
     "\n",
     "Let's say we want to return the $gri$ magnitudes of the top 10 objects in the SMASH DR1 dataset and get it back as a CSV file:"
```
Because the order of rows in the DB table is arbitrary, suggest to replace with "Let's say we want to fetch the g,r,i magnitudes from 10 objects in the SMASH DR1 data set, and retrieve the results as a CSV-formatted string."

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -176,7 +200,7 @@
    ],
    "source": [
-    "query = 'select gmag, rmag, imag from smash_dr1.object limit 10'\n",
+    "query = 'SELECT gmag, rmag, imag FROM smash_dr1.object LIMIT 10'\n",
```
The query is SQL, but the call uses ADQL (it works b/c our system converts on the backend. Also, the query as-is returns many '99.99' values, which are 'invalid' encodings in SMASH. Suggest to change this cell to:

```
query = 'SELECT gmag, rmag, imag FROM smash_dr1.object WHERE gmag < 99 LIMIT 10'
response = qc.query(sql = query, fmt = 'csv')
print (response)
```

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -217,8 +243,7 @@
     "try:\n",
-    "    response = qc.query (adql=query, fmt='csv', \n",
-    "                                  out='vos://zxmags.csv')\n",
+    "    response = qc.query (adql=query, fmt='csv', out='vos://zxmags.csv')\n",
```
Unclear/confusing maybe why the output file is called zxmags?

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -185,10 +209,12 @@
+    "VOSpace is a convenient storage space for users to save their work. It can store any data or file type.  \n",
     "\n",
-    "Now we want to save the results from the same query to our virtual storage space instead.  By putting the query in a try-block we are able to trap errors when executing the query.  Note that running this cell multiple times will trigger an error and we use the Storage Manager client to remove the file once we are done."
+    "Now we want to save the results from the same query to our virtual storage space instead.  By putting the query in a try-block we are able to trap errors when executing the query.  Note that running this cell multiple times will trigger an error and we use the Storage Manager Client to remove the file once we are done."
```
I think this should be maybe in 2-3 cells, and display that after running the query there is a new file in VOSpace?

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -239,9 +264,12 @@
-    "Alternatively we may want to store the results in a table called <i>mags</i> in our remote database."
+    "MyDB is a useful storage space for users to save their work as a table. It can only store data tables.\n",
+    "\n",
+    "Here we show how to store the results in a table called *mags* in our remote database MyDB."
```
These two sentences should be probably combined. I.e. i) state that MyDB os remote per-user relational database, ii) that it can hold tables, iii) and that results of queries can be saved directly to MyDB. Then show example.

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -294,39 +315,420 @@
+    "<a class=\"anchor\" id=\"importmydb\"></a>\n",
+    "# Import data into a MyDB table with QueryClient\n",
+    "\n",
+    "Users can use this function to import data saved on a local computer or import data from VOSpace into a MyDB data table. The data must be in the form of either a CSV file or Pandas Dataframe object in order to load it into MyDB.\n",
```
Unclear at this point in the NB _which_ function is meant.

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -294,39 +315,420 @@
+    }
+   ],
+   "source": [
+    "query = \"select * from gaia_dr1.gaia_source limit 10\"\n",
```
Again this query is SQL but the call is ADQL. Suggest to change the call to `sql=query`

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -294,39 +315,420 @@
+    }
+   ],
+   "source": [
+    "qc.mydb_import('resultname','./gaiaresult.csv')"
```
Wouldn't it be better if the table name in MyDB was more descriptive than 'resultname'?

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -294,39 +315,420 @@
+    }
+   ],
+   "source": [
+    "qc.mydb_import('anotherresultname','vos://newmags.csv')"
```
Same here, the table name could be more descriptive? (it's OK if it contains the string 'test' or 'testing' or similar)

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -294,39 +315,420 @@
+    "<a class=\"anchor\" id=\"async\"></a>\n",
+    "# An asynchronous query\n",
+    "\n",
+    "We now want to run a longer query, say, counting the total number of objects in USNO-B1, and need to do it asynchronously. The first thing we do is submit the query as normal but with the *async_* argument indicated - this will return the id of the asynchronous job. All of the previous arguments can also be used to specify where and in what format we want the query results."
```
So, this section about async queries and result retrieval needs to be adapted so that it can run in auto-testing mode. I.e. use the method
```
qc.query(sql=query,async_=True,wait=True,poll=5)
```
and also explain or link to the "pedestrian" way with getting jobid, checking status until 'COMPLETED', retrieving results with qc.results(). (You can link to the 101 NB for that)

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -339,23 +741,25 @@
+    "# Using profiles\n",
     "\n",
-    "Different datasets (or versions of the same dataset) may reside on different backend servers and a user may want to work explicitly with a (typically older) dataset. In some cases these servers will be used only by developers or those with restricted access.\n",
+    "Different datasets (or versions of the same dataset) may reside on different backend servers and a user may want to work explicitly with a (typically older) dataset. In some cases these servers will be used only by developers or those with restricted access. External TAP services can be accessed by using profiles.\n",
```
When you run `qc.profiles()` right now, no internal profiles are shown (and this is the default case on most days). To make it clearer, and also to combine it with the external profiles sentence, I suggest:

"Sometimes, different datasets ... may reside ... Also, external TAP services are accessible from within Data Lab. In both use cases, the `qc.profiles()` function comes in handy."

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -410,14 +816,410 @@
+    }
+   ],
+   "source": [
+    "listing = sc.ls (name='vos://')\n",
```
Some users have quite a number of files in the VOSpace. It would help to list them in the long format, i.e. I suggest:
```
listing = sc.ls (name='vos://',format='long')
print(listing)
```

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -410,14 +816,410 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can view the data table on-screen by converting it to a pandas dataframe using the *convert* function:"
```
I wouldn't make the case that to view the data on screen one can convert to Pandas. It's more likley user wants to work with the data. Suggest "We can load the data set and e.g. convert to a Pandas data frame, like this:"

rnikutta commented on 04_HowTos/QueryClient/How_to_use_the_Data_Lab_QueryClient.ipynb:
```diff
@@ -436,28 +1238,26 @@
    "metadata": {},
    "source": [
-    "So let's try a query against the default profile first - let's get a list of all tables in the default database.  Note that in this case we are accessing the *information_schema* table of the database, this table is not included in the TAP service and so we <b>must</b> use the *sql* argument to talk directly to the database."
+    "So let's try a query against the default profile first - let's get a list of all tables in the default database.  Note that in this case we are accessing the *information_schema* table of the database, this table is not included in the TAP service and so we **must** use the *sql* argument to talk directly to the database."
```
This example is last, but says "let's use this first". Suggest to either move this example up, or to remove the 'first'.

jacquesalice commented:
> hd5 needs a closing parenthesis
Done.
> Link to the NB directly (on Github)
Done.
> Because the order of rows in the DB table is arbitrary, suggest to replace with "Let's say we want to fetch the g,r,i magnitudes from 10 objects in the SMASH DR1 data set, and retrieve the results as a CSV-formatted string."
Done.
> The query is SQL, but the call uses ADQL (it works b/c our system converts on the backend. Also, the query as-is returns many '99.99' values, which are 'invalid' encodings in SMASH. Suggest to change this cell to:
> 
> ```
> query = 'SELECT gmag, rmag, imag FROM smash_dr1.object WHERE gmag < 99 LIMIT 10'
> response = qc.query(sql = query, fmt = 'csv')
> print (response)
> ```
Done.
> Unclear/confusing maybe why the output file is called zxmags?
That's what it was named before idk but I will change it! Changed it to examplemags
> I think this should be maybe in 2-3 cells, and display that after running the query there is a new file in VOSpace?
I agree. Done.
> These two sentences should be probably combined. I.e. i) state that MyDB os remote per-user relational database, ii) that it can hold tables, iii) and that results of queries can be saved directly to MyDB. Then show example.
Ok how about this: 
"MyDB is a useful OS remote per-user relational database that can store data tables. Furthermore, the results of queries can be directly saved to MyDB, as we show in the following example:"
> Unclear at this point in the NB _which_ function is meant.
Changed to "Users can use the mydb_import function to...."
> Again this query is SQL but the call is ADQL. Suggest to change the call to sql=query
Done.
> Wouldn't it be better if the table name in MyDB was more descriptive than 'resultname'?
> Same here, the table name could be more descriptive? (it's OK if it contains the string 'test' or 'testing' or similar)
I thought it was fun lol but I changed them to `testresult` and `testresult2`
> So, this section about async queries and result retrieval needs to be adapted so that it can run in auto-testing mode. I.e. use the method
> 
> ```
> qc.query(sql=query,async_=True,wait=True,poll=5)
> ```
> 
> and also explain or link to the "pedestrian" way with getting jobid, checking status until 'COMPLETED', retrieving results with qc.results(). (You can link to the 101 NB for that)
Added link to JupyterPythonSQL101 notebook.
> When you run `qc.profiles()` right now, no internal profiles are shown (and this is the default case on most days). To make it clearer, and also to combine it with the external profiles sentence, I suggest:
> 
> "Sometimes, different datasets ... may reside ... Also, external TAP services are accessible from within Data Lab. In both use cases, the `qc.profiles()` function comes in handy."
Done.
> Some users have quite a number of files in the VOSpace. It would help to list them in the long format, i.e. I suggest:
> 
> ```
> listing = sc.ls (name='vos://',format='long')
> print(listing)
> ```
Ahh so that's what that's for! Done.
> I wouldn't make the case that to view the data on screen one can convert to Pandas. It's more likley user wants to work with the data. Suggest "We can load the data set and e.g. convert to a Pandas data frame, like this:"
Done.
> This example is last, but says "let's use this first". Suggest to either move this example up, or to remove the 'first'.
Done.

jacquesalice commented:
> I propose to let this be two cases, both described:
> 
> 1. a manual retrieval using a jobid:
> 
> jobid = qc.query(sql=query,async+=True)
> print(jobid)
> print(qc.status(jobid)) # repeat until 'COMPLETED'
> result = qc.results(jobid)
> 
> 1. A mechanism to run an async query but to make the client auto-loop until the result is available:
> 
> result = qc.query(adql=query,async_=True,wait=True,poll=1,verbose=1)
> 
> Only case 2) should be coded in the notebook. (but case 1 can be linked to in the
> 01_GettingStartedWithDataLab/01_JupyterPythonSQL101.ipynb NB, or even repeated here).

Done.
